compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs

```diff
@@ -127,9 +137,16 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 let did = did.expect_local();
                 let hir_id = self.infcx.tcx.hir().local_def_id_to_hir_id(did);
 
-                if let Some((span, name)) =
+                if let Some((span, place)) =
                     self.infcx.tcx.typeck(did).closure_kind_origins().get(hir_id)
                 {
+                    let name = match place.base {
```

Comment:
We also want to print precise paths, so `p.x` was used etc. I could've probably made more ephasis on this in the issue 😅

A good place to look to figure out the how, is  probably `rustc_mir/src/borrow_check`.